### PR TITLE
added support for azure native fencing

### PR DESCRIPTION
**Pull request by mr-stringer**
An extensive set of changes to add the option to use the Azure native fencing method.

### CODE REVIEW

**Review by mr-stringer — COMMENTED** [review]
I've added comments to the changes in the files, please take a look at them.

File: .gitignore [hunk 1]
```diff
@@ -7,6 +7,7 @@ terraform/aws/terraform.tfvars
 terraform/azure/.terraform
 terraform/azure/terraform.tfstate*
 terraform/azure/terraform.tfvars
+terraform/azure/fence_data.json
```
Comment (mr-stringer) [inline .gitignore:10]: When using the azure fence agent, we need additional information when creating the cluster.  This information is known to terraform but not Ansible.  Therefore the file ./terraform/azure/fence_data.json is created.  When using azure native fencing, this file can be sourced to get the sub id, tenant id and resource group.  It's added to .gitignore as it contains sensitive data.

File: ansible/playbooks/sap-hana-cluster.yaml [hunk 2]
```diff
@@ -5,6 +5,13 @@
   become_user: root
   vars:
     provider: azure # only provider for now
+    use_sbd: yes
```
Comment (mr-stringer) [inline ansible/playbooks/sap-hana-cluster.yaml:8]: All the vars required for Azure fencing are added to the top of the playbook as empty.  The glue code will need to override these variables to use Azure fencing.  SBD remains the default method.

File: ansible/playbooks/sap-hana-cluster.yaml [hunk 3]
```diff
@@ -13,7 +20,7 @@
 
   tasks:
 
-  - name: SBD Cluster [azure]
+  - name: Base Cluster [azure]
```
Comment (mr-stringer) [inline ansible/playbooks/sap-hana-cluster.yaml:23]: Renamed as this is base clustering for both SBD and Azure fencing

File: ansible/playbooks/tasks/azure-cluster-bootstrap.yaml [hunk 4]
```diff
@@ -9,10 +9,46 @@
...
+- name: Ensure Azure Python SDK and Azure Identity python modules are installed
+  community.general.zypper:
+    name: "{{ item.package }}"
+    state: present
+  loop:
+    # For SLES 12 a we need to force a downgrade of python-azure-core see https://www.suse.com/support/kb/doc/?id=000020716
+    - { 'sles' : '12', 'package' : ['python-azure-mgmt-compute', 'python-azure-identity', 'python-azure-core==1.9.0-2.3.4'] } 
+    - { 'sles' : '15', 'package' : ['python3-azure-mgmt-compute', 'python3-azure-identity'] }
+  when: item.sles == ansible_distribution_major_version
+
```
Comment (mr-stringer) [inline ansible/playbooks/tasks/azure-cluster-bootstrap.yaml:51]: Some additional set-up actions.  Some are for both scenarios, some are just for Azure fencing.

File: ansible/playbooks/tasks/azure-cluster-bootstrap.yaml [hunk 5]
```diff
@@ -177,17 +235,17 @@
...
     cln_azure_events: "{{ crm_conf_show.stdout | regex_search('clone cln_azure-events') }}"
   when: ansible_hostname == 'vmhana01'
 
-- name: Ensure maintainence mode is deactivated
+- name: Ensure maintenance mode is deactivated
   ansible.builtin.command: crm configure property maintenance-mode=false
   when:
     - ansible_hostname == 'vmhana01'
-    - crm_maintainence_mode is true or crm_maintainence_mode == 'unknown'
+    - crm_maintenance_mode is true or crm_maintenance_mode == 'unknown'
```
Comment (mr-stringer) [inline ansible/playbooks/tasks/azure-cluster-bootstrap.yaml:251]: Coded added to skip certain steps depending on the fencing method used.  Also fixed the spelling of `maintenance ` in several vars :)

File: ansible/playbooks/tasks/azure-cluster-hana.yaml [hunk 6]
```diff
@@ -29,6 +31,22 @@
...
+  ansible.builtin.command: crm configure property stonith-timeout=900
+  when:
+    - stonith_timeout | length == 0
+    - use_sbd | bool
+
+- name: Enable stonith
+  ansible.builtin.command: crm configure property stonith-enabled=true
+  when: stonith_enabled != 'true'
+
 - name: Create HANA topology clone
```
Comment (mr-stringer) [inline ansible/playbooks/tasks/azure-cluster-hana.yaml:50]: Ensures that the rebuild of the SBD STONITH resources only occurs when using SBD

File: ansible/playbooks/vars/hana_hsr_parameters.yaml [hunk 7]
```diff
@@ -1,12 +1,7 @@
 sap_hana_sid: "HDB"
 sap_hana_instance_number: "00"
 sap_hana_install_master_password: "EasyPassw0rd!"
-
-### Cluster Definition
-sap_ha_install_pacemaker_cluster_name: hdbcluster
-sap_hana_hacluster_password: "hdbclusterPassw0rd"
-
-sap_domain: qe-test.suse.com
+sap_domain: "qe-test.suse.com"
```
Comment (mr-stringer) [inline ansible/playbooks/vars/hana_hsr_parameters.yaml:4]: Removed superfluous variables.

File: requirements.txt [hunk 8]
```diff
@@ -1,3 +1,4 @@
 ansible==4.10.0
 ansible-core==2.11.12
-awscli==1.19.48
\ No newline at end of file
+awscli==1.19.48
+jmespath==1.0.1
```
Comment (mr-stringer) [inline requirements.txt:4]: Added jmespath to the requirements file as this is required by ansible to decode the fence_data.json file

File: terraform/azure/inventory.tmpl [hunk 9]
```diff
@@ -5,15 +5,15 @@ all:
...
  %{ for index, value in iscsi-pip ~}
         ${iscsi-name[index]}:
           ansible_host: ${value}
-          ansible_python_interpreter: /usr/bin/python3
+          %{ if iscsi-major-version == "12" }ansible_python_interpreter: /usr/bin/python2.7%{ else }ansible_python_interpreter: /usr/bin/python3%{ endif }
 %{ endfor ~}
 %{ endif }
-  hosts: null
\ No newline at end of file
+  hosts: null
```
Comment (mr-stringer) [inline terraform/azure/inventory.tmpl:19]: ansible_python_path is now based on the os_image used or by a specific override in tfvars.  SLES 12 will use python2.7 and 15 will use python3

File: terraform/azure/main.tf [hunk 10]
```diff
@@ -50,6 +50,16 @@ locals {
...
+  hana_major_version       = var.hana_os_major_version != "" ? var.hana_os_major_version : regex("[a-zA-Z]+:[a-zA-Z]+-[a-zA-Z]+-([0-9]+)-", local.hana_os_image)[0]
+  iscsi_major_version      = var.iscsi_os_major_version != "" ? var.iscsi_os_major_version : regex("[a-zA-Z]+:[a-zA-Z]+-[a-zA-Z]+-([0-9]+)-", local.iscsi_os_image)[0]
+  monitoring_major_version = var.monitoring_os_major_version != "" ? var.monitoring_os_major_version : regex("[a-zA-Z]+:[a-zA-Z]+-[a-zA-Z]+-([0-9]+)-", local.monitoring_os_image)[0]
+  drbd_major_version       = var.drdb_os_major_version != "" ? var.drdb_os_major_version : regex("[a-zA-Z]+:[a-zA-Z]+-[a-zA-Z]+-([0-9]+)-", local.drbd_os_image)[0]
+  netweaver_major_version  = var.netweaver_os_major_version != "" ? var.netweaver_os_major_version : regex("[a-zA-Z]+:[a-zA-Z]+-[a-zA-Z]+-([0-9]+)-", local.netweaver_os_image)[0]
+  bastion_major_version    = var.bastion_os_major_version != "" ? var.bastion_os_major_version : regex("[a-zA-Z]+:[a-zA-Z]+-[a-zA-Z]+-([0-9]+)-", local.bastion_os_image)[0]
+
   # Netweaver password checking
```
Comment (mr-stringer) [inline terraform/azure/main.tf:63]: New locals to allow ansible_python_path to be based on os_image

File: terraform/azure/outputs.tf [hunk 11]
```diff
@@ -5,6 +5,11 @@
 # - Public node name
 
 # iSCSI server
+output "hana_os_major_verion" {
+  description = "The major version of the HANA OS"
+  value       = local.hana_major_version
+}
+
```
Comment (mr-stringer) [inline terraform/azure/outputs.tf:13]: If the os_image doesn't match the conventional pattern ("SUSE:sles-sap-15-sp3:gen2:latest"), the major version can be specified with these x_os_major_version.  This is used to set the ansible_python_path

**Review by lpalovsky — APPROVED** [review]
LGTM

**Review by alvarocarvajald — APPROVED** [review]
LGTM

File: terraform/azure/inventory.tmpl (hunk 9 above)
Comment (mpagot) [reply terraform/azure/inventory.tmpl:19]: Quite like, needs to be ported to other Cloud providers templates too, right?

Comment (alvarocarvajald) [reply terraform/azure/inventory.tmpl:19]: Yes. This is only for Azure at the moment

**Pull request merged**